fsimpl/unionfs/unionfs.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionfs
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/hugelgupf/p9/fsimpl/readdir"
+	"github.com/hugelgupf/p9/fsimpl/templatefs"
+	"github.com/hugelgupf/p9/linux"
+	"github.com/hugelgupf/p9/p9"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	ErrFlatHierarchy = errors.New("unionfs only supports a flat hierarchy")
+	ErrFileExists    = errors.New("file already exists")
+)
+
+type Opt func(fs *FS) error
+
+// FS is a p9.Attacher.
+type FS struct {
+	mounts map[string]p9.File
+}
+
+func WithMount(dir string, attacher p9.Attacher) Opt {
+	return func(fs *FS) error {
+		if strings.Contains(dir, string(filepath.Separator)) {
+			return fmt.Errorf("%w: directories are not supported: %s", ErrFlatHierarchy, dir)
+		}
+		if _, ok := fs.mounts[dir]; ok {
+			return fmt.Errorf("%w: %s", ErrFileExists, dir)
+		}
+
+		f, err := attacher.Attach()
+		if err != nil {
+			return err
+		}
+		fs.mounts[dir] = f
+		return nil
+	}
+}
+
+func WithDir(dir string, mounts ...Opt) Opt {
+	return func(fs *FS) error {
+		subfs, err := New(mounts...)
+		if err != nil {
+			return err
+		}
+
+		return WithMount(dir, subfs)(fs)
+	}
+}
+
+func WithFile(file string, f p9.File) Opt {
```

Comment:
actually, "thing exists" may not matter. e.g., in Plan 9, you could create a pattern which was the same as the existing pattern, and direct that it go before, after, or replace (MBEFORE, MAFTER, MREPL) the existing one. If we were to allow this, we could do stuff like this:
/bin -> to /some/where/bin
/bin -> to /home/you/bin
/home -> to /server1/home
/home -> to /server2/home

it would mean the mounts was not a map, but a list, but ... it's not like the map will really save time over the list, I expect, as these mount tables are generally small (16 elements is a big one)

Then, on mount traversal, if you were walking to /home/a, and it did not exist in the first bind, and it did exist in the second one, that would be ok. 
This is how Plan 9 constructed /bin as a union of multiple other /bin. 
It's been just long enough since I did this that I'm forgetting exact syntax, but I can look it up, if we think we might want it.

but, e.g, 
bind /tmp/cpu/bin /bin
bind -a /tmp/cpu//home/rminnich/bin /bin
would map to a union mount
/bin ->  /tmp/cpu/bin
/bin -> /tmp/cpu/home/rminnich/bin

or I might put my bin first, so I can have my versions of things first.

This is why Plan 9 did not need $PATH.